Lib/functools.py

```diff
@@ -949,6 +952,11 @@ def _method(*args, **kwargs):
         _method.__isabstractmethod__ = self.__isabstractmethod__
         _method.register = self.register
         update_wrapper(_method, self.func)
+
+        # not all objects have __dict__ (e.g. classes with __slots__)
+        if hasattr(obj, '__dict__'):
+            obj.__dict__[self.attrname] = _method
```

Comment:
Don't we have to handle the self.attrname is None case?
For example, cached_property handle this case
https://github.com/python/cpython/blob/2cf99026d6320f38937257da1ab014fc873a11a6/Lib/functools.py#L983-L985

Reply:
@corona10 is it possible to have an instance of `singledispatchmethod` that hasn't had `__set_name__` called on it?

Reply:
> @corona10 is it possible to have an instance of singledispatchmethod in the dictionary of a created class that hasn't had __set_name__ called on it? 

If it is impossible, I would like to suggest adding an assert statement as the invariant condition.

Reply:
> > @corona10 is it possible to have an instance of singledispatchmethod in the dictionary of a created class that hasn't had **set_name** called on it?
> 
> If it is impossible, I would like to suggest adding an assert statement as the invariant condition.

None of the examples of descriptors with `__set_name__` in [the Descriptor HOWTO](https://docs.python.org/3/howto/descriptor.html#customized-names) have assert statements like that. It feels unnecessary to me, given that https://peps.python.org/pep-0487/#implementation-details specifies that `__set_name__` must be called on all descriptors in a class namespace immediately after class creation. And I worry that any extra assertions in this method will slow it down, which is what this PR is trying to avoid :-)

Reply:
Well, okay, then let's leave a comment about the fact instead of assert statement :) 
I don't think that all people understand the mechanism of PEP-0487. It will be helpful to future contributors or future reviewers.

Reply:
A comment sounds good to me!

Reply:
> I think that branch in cached_property.__get__ might be overly cautious. I'm not sure it's actually reachable. No tests fail if I apply this diff to functools.py:

And will you submit the patch too? It looks good to be removed.

Reply:
> > I think that branch in cached_property.**get** might be overly cautious. I'm not sure it's actually reachable. No tests fail if I apply this diff to functools.py:
> 
> And will you submit the patch too? It looks good to be removed.

I'll open an issue and see what Carl Meyer thinks about it, since he's the expert on `cached_property` :)